test/triangular_test.js

```diff
@@ -0,0 +1,255 @@
+        },
+    );
+  });
+
+  it('triangular diagonal=-1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            0, 6, 3,
+          ],
+        },
+        {
+          diagonal: -1,
+        },
+    );
+  });
+
+  it('triangular fully copied diagonal=-4', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          diagonal: -4,
+        },
+    );
+  });
+
+  it('triangular upper=false', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 0, 0,
+            9, 4, 0,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+        },
+    );
+  });
+
+  it('triangular upper=false diagonal=1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 0,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: 1,
+        },
+    );
+  });
+
+  it('triangular fully copied upper=false diagonal=4 ', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: 4,
+        },
+    );
+  });
+
+  it('triangular upper=false diagonal=-1', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
+          data: [
+            0, 0, 0,
+            9, 0, 0,
+            2, 6, 0,
+          ],
+        },
+        {
+          upper: false,
+          diagonal: -1,
+        },
+    );
+  });
+
+  it('triangular fully zero upper=false diagonal=-4', function() {
+    testTriangular(
+        {
+          shape: [3, 3],
+          data: [
+            7, 1, 2,
+            9, 4, 8,
+            2, 6, 3,
+          ],
+        },
+        {
+          shape: [3, 3],
```

Comment:
We should also include one with a batch count other than 0. 🤨 It looks like a spec typo that I missed where it says `triangular` takes exactly 2D tensors, but instead it should say the input is *at least* 2D. You can see other ML libraries just treat each 2D subsection as a separate batch:

https://numpy.org/doc/stable/reference/generated/numpy.triu.html
https://pytorch.org/docs/stable/generated/torch.tril.html
https://onnx.ai/onnx/operators/onnx__Trilu.html

So shape [4,3,5,7] would apply the triangular masking to the last two dimensions. Let's create a WebNN issue...

Reply:
> Let's create a WebNN issue...

Done, please see https://github.com/webmachinelearning/webnn/issues/494, thanks.